MGDSTRM-3724 Add the status code to the keycloak_request_duration_bucket metric

CathalOConnorRH opened the pull request:
## Motivation
https://issues.redhat.com/browse/MGDSTRM-3724
https://issues.redhat.com/browse/MGDSTRM-4318

## What
Add status code and uri to the keycloak_request_duration_bucket metric

## Why
This will allow filtering on specific statuses 
Currently no status are returned in the metric
`keycloak_request_duration_bucket{method="POST",resource="realms,realms/master/protocol/openid-connect",le="50.0",} 0.0
keycloak_request_duration_bucket{method="POST",resource="realms,realms/master/protocol/openid-connect",le="100.0",} 0.0
keycloak_request_duration_bucket{method="POST",resource="realms,realms/master/protocol/openid-connect",le="250.0",} 0.0
keycloak_request_duration_bucket{method="POST",resource="realms,realms/master/protocol/openid-connect",le="500.0",} 0.0
keycloak_request_duration_bucket{method="POST",resource="realms,realms/master/protocol/openid-connect",le="1000.0",} 1.0
keycloak_request_duration_bucket{method="POST",resource="realms,realms/master/protocol/openid-connect",le="2000.0",} 1.0
keycloak_request_duration_bucket{method="POST",resource="realms,realms/master/protocol/openid-connect",le="10000.0",} 1.0
keycloak_request_duration_bucket{method="POST",resource="realms,realms/master/protocol/openid-connect",le="30000.0",} 1.0
keycloak_request_duration_bucket{method="POST",resource="realms,realms/master/protocol/openid-connect",le="+Inf",} 1.0
keycloak_request_duration_count{method="POST",resource="realms,realms/master/protocol/openid-connect",} 1.0
keycloak_request_duration_sum{method="POST",resource="realms,realms/master/protocol/openid-connect",} 625.0`

## How
added status code and uri to the to the bucket output. 

## Verification Steps

# Option 1
1. Build the code
2. Copy the artifact into $KC/standalone/deployments
3. Make sure the metrics logger is turned on
4. Play with Keycloak a bit
5. Look at the following url: "$KC/auth/realms/master/metrics"

# Option 2
update keycloak cr to use the jar built and released from my fork 
`https://github.com/CathalOConnorRH/keycloak-metrics-spi/releases/tag/v2.4.1-beta`


Metrics should look similar to 
`keycloak_request_duration_bucket{code="200",method="POST",resource="realms,realms/master/protocol/openid-connect",uri="realms/master/protocol/openid-connect/token",le="50.0",} 0.0
keycloak_request_duration_bucket{code="200",method="POST",resource="realms,realms/master/protocol/openid-connect",uri="realms/master/protocol/openid-connect/token",le="100.0",} 0.0
keycloak_request_duration_bucket{code="200",method="POST",resource="realms,realms/master/protocol/openid-connect",uri="realms/master/protocol/openid-connect/token",le="250.0",} 1.0
keycloak_request_duration_bucket{code="200",method="POST",resource="realms,realms/master/protocol/openid-connect",uri="realms/master/protocol/openid-connect/token",le="500.0",} 1.0
keycloak_request_duration_bucket{code="200",method="POST",resource="realms,realms/master/protocol/openid-connect",uri="realms/master/protocol/openid-connect/token",le="1000.0",} 1.0
keycloak_request_duration_bucket{code="200",method="POST",resource="realms,realms/master/protocol/openid-connect",uri="realms/master/protocol/openid-connect/token",le="2000.0",} 1.0
keycloak_request_duration_bucket{code="200",method="POST",resource="realms,realms/master/protocol/openid-connect",uri="realms/master/protocol/openid-connect/token",le="10000.0",} 1.0
keycloak_request_duration_bucket{code="200",method="POST",resource="realms,realms/master/protocol/openid-connect",uri="realms/master/protocol/openid-connect/token",le="30000.0",} 1.0
keycloak_request_duration_bucket{code="200",method="POST",resource="realms,realms/master/protocol/openid-connect",uri="realms/master/protocol/openid-connect/token",le="+Inf",} 1.0
keycloak_request_duration_count{code="200",method="POST",resource="realms,realms/master/protocol/openid-connect",uri="realms/master/protocol/openid-connect/token",} 1.0`

## Checklist:

- [x] Code has been tested locally by PR requester
- [ ] Changes have been successfully verified by another team member 

## Progress

- [x] Finished task
- [ ] TODO

## Additional Notes

PS.: Add images and/or .gifs to illustrate what was changed if this pull request modifies the appearance/output of something presented to the users.

Rajagopalan-Ranganathan commented on src/test/java/org/jboss/aerogear/keycloak/metrics/PrometheusExporterTest.java:
```diff
@@ -265,28 +265,28 @@ public void shouldCorrectlyRecordGenericAdminEvents() throws IOException {
     @Test
     public void shouldCorrectlyRecordResponseDurations() throws IOException {
-        PrometheusExporter.instance().recordRequestDuration(5, "GET", "admin,admin/serverinfo");
+        PrometheusExporter.instance().recordRequestDuration(200, 5, "GET", "admin,admin/serverinfo", "auth/realm");
```
A doubt : should the URI be only "auth/realm" or it can be other paths as well? (since all the tests use this , wanted to understand)

CathalOConnorRH replied:
the `auth/realm` URI is just an example as part of the test, Any URI path can be used

Rajagopalan-Ranganathan replied:
ok got it. thanks!

maleck13 replied:
I think it would be good to wait for @pb82. As I understand it just using the RAW URI caused issues in the past especially when there were unique ids in the path as it created too many timeseries. Ideally these URIs would be translated into something like ``` realms/{realm}/protocol/openid-connect/token ``` however that could be a lot of effort so perhaps better to choose some important APIs and just covers those initially.

maleck13 replied:
The openid-connect ones would be important ones along with service account mgmt

pb82 replied:
@CathalOConnorRH the raw URL should not be added as a label, this is what caused https://github.com/aerogear/keycloak-metrics-spi/issues/64 in the past

pb82 replied:
@CathalOConnorRH but it looks like you're already using the resource extractor instead of the raw URL. That's better, but keep in mind that every combination of labels leads to a new exported histogram (and that means one additional mertric per bucket plus sum and total). So if we keep adding labels to histograms the amount of metrics will grow (and has grown to huge numbers in the past).

Rajagopalan-Ranganathan commented:
/lgtm

akoserwal commented:
@pb82 Would like to take a look as well?

akoserwal commented:
LGTM

pb82 commented:
@CathalOConnorRH if it is absolutely required to add those labels to the histogram, can we make that opt-in behind a flag?

CathalOConnorRH commented:
@pb82 I've updated the `ResourceExtractor` to include a `URI_METRICS_ENABLED` boolean and a check for /token URI. 
an example of the metrics output is attached 

Let me know what you think 

[KeycloakMetrics.txt](https://github.com/aerogear/keycloak-metrics-spi/files/6954266/KeycloakMetrics.txt)

CathalOConnorRH commented:
Hey @pb82 Have you had any thoughts on the above changes ?

pb82 commented:
thanks @CathalOConnorRH I'll give it a final try and then merge

Rajagopalan-Ranganathan commented:
@pb82 did you find time to try it out? Just curious about the status. Thanks!

maleck13 commented on src/main/java/org/jboss/aerogear/keycloak/metrics/ResourceExtractor.java:
```diff
@@ -54,4 +54,22 @@ static String getResource(UriInfo uriInfo) {
+     * @return The resource uri.
+     */
+    static String getURI(UriInfo uriInfo) {
+        if (URI_METRICS_ENABLED) {
```
does it make sense here to replace certain parts of the URI with a place holder to reduce uniqueness? For example replace ``` /realm/master ``` with ```/realm/{realm}``` for example. Looking for ways to reduce uniqueness while still providing the needed insight

CathalOConnorRH replied:
That's possible, We could do another flag to enable more detailed realm output if needed ?

pb82 replied:
@CathalOConnorRH if we add that, we can also add the change were we remove the labels completely if URI_METRICS_ENABLED is not set (at the moment they are put on all metrics with an empty value).

CathalOConnorRH replied:
agreed, I'll work on both of those changes today

Rajagopalan-Ranganathan commented on src/main/java/org/jboss/aerogear/keycloak/metrics/MetricsFilter.java:
```diff
@@ -23,6 +23,8 @@
     private static final String METRICS_REQUEST_TIMESTAMP = "metrics.requestTimestamp";
     private static final MetricsFilter INSTANCE = new MetricsFilter();
 
+    private static final boolean URI_METRICS_ENABLED = Boolean.getBoolean("URI_METRICS_ENABLED");
```
Not a JAva guy, so trying to understand would it default to false, in case the variable is not defined?

pb82 replied:
according to the docs:

```
Returns true if and only if the system property named by the argument exists and is equal to the string "true". (Beginning with version 1.0.2 of the JavaTM platform, the test of this string is case insensitive.) A system property is accessible through getProperty, a method defined by the System class.

If there is no property with the specified name, or if the specified name is empty or null, then false is returned.
```

CathalOConnorRH replied:
Yeah, if the value is set and set to true it will return true, otherwise it returns false 
https://www.tutorialspoint.com/java/lang/boolean_getboolean.htm

CathalOConnorRH commented:
@pb82  I've implemented the detailed output and removed the uri when it's not enabled. 
Some sample output of each of the three scenarios are attached. 
Let me know what you think  
[URIEnabledURIDetailed.txt](https://github.com/aerogear/keycloak-metrics-spi/files/7082397/URIEnabledURIDetailed.txt)
[URIDisabled.txt](https://github.com/aerogear/keycloak-metrics-spi/files/7082399/URIDisabled.txt)
[URIEnabled.txt](https://github.com/aerogear/keycloak-metrics-spi/files/7082401/URIEnabled.txt)

Rajagopalan-Ranganathan commented:
@pb82 We need this change in MAS SSO soon, to build some metrics - dashboards and also to investigate and get precise information regarding our load (which end points are used most). Can you take a look and close this at the earliest?
